backend-api/src/functions/adapters/dynamoDbAdapter.ts

```diff
@@ -0,0 +1,195 @@
+import {
+  AttributeValue,
+  ConditionalCheckFailedException,
+  DynamoDBClient,
+  PutItemCommand,
+  PutItemCommandInput,
+  QueryCommand,
+  QueryCommandInput,
+  QueryCommandOutput,
+} from "@aws-sdk/client-dynamodb";
+import { errorResult, Result, successResult } from "../utils/result";
+import {
+  CreateSessionAttributes,
+  SessionDetails,
+} from "../services/sessionService/sessionService";
+import { NodeHttpHandler } from "@smithy/node-http-handler";
+
+type DynamoDbRecord = Record<string, AttributeValue>;
+
+export class DynamoDbAdapter {
+  private readonly tableName: string;
+  private readonly dynamoDbClient = new DynamoDBClient({
+    region: process.env.REGION,
+    maxAttempts: 2,
+    requestHandler: new NodeHttpHandler({
+      connectionTimeout: 29000,
+      requestTimeout: 29000,
+    }),
+  });
+
+  constructor(tableName: string) {
+    this.tableName = tableName;
+  }
+
+  async readSessionId(
+    subjectIdentifier: string,
+  ): Promise<Result<string | null>> {
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    const input: QueryCommandInput = {
+      TableName: this.tableName,
+      IndexName: "subjectIdentifier-timeToLive-index",
+      KeyConditionExpression:
+        "subjectIdentifier = :subjectIdentifier and :currentTimeInSeconds < timeToLive",
+      FilterExpression: "sessionState = :sessionState",
+      ExpressionAttributeValues: {
+        ":subjectIdentifier": { S: subjectIdentifier },
+        ":sessionState": { S: "ASYNC_AUTH_SESSION_CREATED" },
```

Comment:
Thinking ahead a bit here - we're going to soon start getting the session record when looking up via the sessionId

I'm keen not to hardcode this string as we'll soon forget it's hardcoded.

Perhaps we should have some form of constants file in the sessionService folder?

const sessionStates = {
    ASYNC_AUTH_SESSION_CREATED: "ASYNC_AUTH_SESSION_CREATED"
} as const

or something similar

The same for when it's used in the next function

Reply:
Done but I did not put it in a separate file - it's within DynamoDB adapter. Perhaps it should be in SessionService?

Reply:
hmmm I'm torn

Reply:
If we update the DynamoDBAdapter file name to include the word Session I'd love for it to be in that file

Reply:
Let's keep it in the DynamoDB file as that's the only file that I think should know about this. If we find in other lambdas that they need access to the authSessionState we can relocate!